pandas/core/reshape/merge.py

```diff
@@ -758,12 +758,31 @@ def _get_join_info(self):
             if self.right_index:
                 if len(self.left) > 0:
                     join_index = self.left.index.take(left_indexer)
+                    if (self.how == 'right' and -1 in left_indexer
+                            and not isinstance(self.right.index, MultiIndex)):
```

Comment:
make this into a function, something like
```
def create_join_index(index, indexer, other_indexer):
   ....
```

I think you can *always* call this, just not doing the replacement if what you are calling ``absent`` is empty